### DISCUSSION TITLE
Self hosted instructions doesn't seem correct

### DISCUSSION THREAD

**Discussion by kenshinjeff** [supabase/supabase#11600] [Questions] [answered]
Hi everyone, I followed the instructions from the official docs, and I'm trying to get the demo up without editing the .env file. But I'm encountering these error messages from realtime-dev.supabase-realtime. How do I go about fixing this? 

Install reference: https://supabase.com/docs/guides/self-hosting/docker 
Ubuntu 22.04 x64 on Vultr's High Frequency plan
Docker version 20.10.17, build 100c70180f 

> realtime-dev.supabase-realtime  | 00:48:57.084 [error] Could not create schema migrations table. This error usually happens due to the following:
> realtime-dev.supabase-realtime  |
> realtime-dev.supabase-realtime  |   * The database does not exist
> realtime-dev.supabase-realtime  |   * The "schema_migrations" table, which Ecto uses for managing
> realtime-dev.supabase-realtime  |     migrations, was defined by another library
> realtime-dev.supabase-realtime  |   * There is a deadlock while migrating (such as using concurrent
> realtime-dev.supabase-realtime  |     indexes with a migration_lock)
> realtime-dev.supabase-realtime  |
> realtime-dev.supabase-realtime  | To fix the first issue, run "mix ecto.create".
> realtime-dev.supabase-realtime  |
> realtime-dev.supabase-realtime  | To address the second, you can run "mix ecto.drop" followed by
> realtime-dev.supabase-realtime  | "mix ecto.create". Alternatively you may configure Ecto to use
> realtime-dev.supabase-realtime  | another table and/or repository for managing migrations:
> realtime-dev.supabase-realtime  |
> realtime-dev.supabase-realtime  |     config :realtime, Realtime.Repo,
> realtime-dev.supabase-realtime  |       migration_source: "some_other_table_for_schema_migrations",
> realtime-dev.supabase-realtime  |       migration_repo: AnotherRepoForSchemaMigrations
> realtime-dev.supabase-realtime  |
> realtime-dev.supabase-realtime  | The full error report is shown below.
> realtime-dev.supabase-realtime  |
> realtime-dev.supabase-realtime  | 00:48:57.087 [error] Postgrex.Protocol (#PID<0.136.0>) failed to connect: ** (DBConnection.ConnectionError) tcp connect (db:5432): timeout
> realtime-dev.supabase-realtime  | 00:48:57.087 [error] Postgrex.Protocol (#PID<0.135.0>) failed to connect: ** (DBConnection.ConnectionError) tcp connect (db:5432): timeout
> realtime-dev.supabase-realtime  | ** (DBConnection.ConnectionError) connection not available and request was dropped from queue after 14978ms. This means requests are coming in and your connection pool cannot serve them fast enough. You can address this by:
> realtime-dev.supabase-realtime  |
> realtime-dev.supabase-realtime  |   1. Ensuring your database is available and that you can connect to it
> realtime-dev.supabase-realtime  |   2. Tracking down slow queries and making sure they are running fast enough
> realtime-dev.supabase-realtime  |   3. Increasing the pool_size (although this increases resource consumption)
> realtime-dev.supabase-realtime  |   4. Allowing requests to wait longer by increasing :queue_target and :queue_interval
> realtime-dev.supabase-realtime  |
> realtime-dev.supabase-realtime  | See DBConnection.start_link/2 for more information
> realtime-dev.supabase-realtime  |
> realtime-dev.supabase-realtime  |     (ecto_sql 3.8.3) lib/ecto/adapters/sql.ex:932: Ecto.Adapters.SQL.raise_sql_call_error/1
> realtime-dev.supabase-realtime  |     (elixir 1.14.0) lib/enum.ex:1658: Enum."-map/2-lists^map/1-0-"/2
> realtime-dev.supabase-realtime  |     (ecto_sql 3.8.3) lib/ecto/adapters/sql.ex:1024: Ecto.Adapters.SQL.execute_ddl/4
> realtime-dev.supabase-realtime  |     (ecto_sql 3.8.3) lib/ecto/migrator.ex:696: Ecto.Migrator.verbose_schema_migration/3
> realtime-dev.supabase-realtime  |     (ecto_sql 3.8.3) lib/ecto/migrator.ex:510: Ecto.Migrator.lock_for_migrations/4
> realtime-dev.supabase-realtime  |     (ecto_sql 3.8.3) lib/ecto/migrator.ex:422: Ecto.Migrator.run/4
> realtime-dev.supabase-realtime  |     (ecto_sql 3.8.3) lib/ecto/migrator.ex:146: Ecto.Migrator.with_repo/3
> realtime-dev.supabase-realtime  |     nofile:1: (file)
> realtime-dev.supabase-realtime exited with code 1

**Comment by w3b6x9** [comment, marked as answer]
@kenshinjeff there was an issue with the self-hosting b/c the db user didn't have the proper permissions. It should have been fixed by this change: https://github.com/supabase/supabase/commit/00ac93d2ee0529bb264f3acc6a0cb042adbfae60

**Reply by kenshinjeff** [reply]
Hi there, it's a bit weird but I have since reinstalled and it starts up fine now. 

sudo docker ps tells me that supabase/storage-api and supabase/gotrue are unhealthy. Is this normal behaviour? 

Accessing the app gives the it's stuck at "Connecting to default project". How should I go about debugging this?

**Reply by kenshinjeff** [reply]
@w3b6x9 You know what, restarting docker gives the same Ecto migrations error I have previously encountered. I'll just move on from trying this out and come back to it at a later time to see if it fits. Thank you for your help so far, I really appreciate it.

**Reply by w3b6x9** [reply]
@kenshinjeff feel free to submit a ticket at https://app.supabase.com/support/new when you want to come back to this. We would love to help you resolve this.

**Reply by Lubjan** [reply]
**edit:** supabase seems to work perfectly fine when `ufw` is disabled, but then it isn't accessible from the outside anymore.
**edit2:** yeah, was all just firewall issues with docker

@w3b6x9 I've just setup supabase on my server and am seeing the same issue here for realtime and storage, your commit for this problem is also present, but still doesn't seem to work.

When checking the `volumes/db/data/` directory I've noticed that nothing seems to be owned by an actual user, but rather `systemd-coredump`, not sure if that has anything to do with it.
I tried changing the owner, but after the next `docker compose up` it reverts the user/group back to `systemd-coredump`, also tried adding `user: "supabase:supabase"` to db in the docker-compose.yml as that was a solution from googling, but didn't help either. 

(`supabase-rest` sometimes also complains about a failed db connection, not sure if that is normal though)

**volumes/db/data/**
![image](https://user-images.githubusercontent.com/16855427/217324814-71aa6e89-a075-4938-a09c-6bd64a5b5d3c.png)

**docker ps**
![image](https://user-images.githubusercontent.com/16855427/217327818-a902d374-708a-4cc3-b102-8fd52149b5dc.png)

**Rest Error**
```log
supabase-rest                   | 07/Feb/2023:17:54:25 +0000: {"code":"PGRST000","details":"connection to server at \"db\" (IP), port 5432 failed: Operation timed out\n\tIs the server running on that host and accepting TCP/IP connections?\n","hint":null,"message":"Database connection error. Retrying the connection."}
supabase-rest                   | 07/Feb/2023:17:54:25 +0000: Attempting to reconnect to the database in 0 seconds...
```


**Relatime and Storage Error**
```log
realtime-dev.supabase-realtime  | 17:36:28.012 [error] Could not create schema migrations table. This error usually happens due to the following:
realtime-dev.supabase-realtime  |
realtime-dev.supabase-realtime  |   * The database does not exist
realtime-dev.supabase-realtime  |   * The "schema_migrations" table, which Ecto uses for managing
realtime-dev.supabase-realtime  |     migrations, was defined by another library
realtime-dev.supabase-realtime  |   * There is a deadlock while migrating (such as using concurrent
realtime-dev.supabase-realtime  |     indexes with a migration_lock)
realtime-dev.supabase-realtime  |
realtime-dev.supabase-realtime  | To fix the first issue, run "mix ecto.create".
realtime-dev.supabase-realtime  |
realtime-dev.supabase-realtime  | To address the second, you can run "mix ecto.drop" followed by
realtime-dev.supabase-realtime  | "mix ecto.create". Alternatively you may configure Ecto to use
realtime-dev.supabase-realtime  | another table and/or repository for managing migrations:
realtime-dev.supabase-realtime  |
realtime-dev.supabase-realtime  |     config :realtime, Realtime.Repo,
realtime-dev.supabase-realtime  |       migration_source: "some_other_table_for_schema_migrations",
realtime-dev.supabase-realtime  |       migration_repo: AnotherRepoForSchemaMigrations
realtime-dev.supabase-realtime  |
realtime-dev.supabase-realtime  | The full error report is shown below.
realtime-dev.supabase-realtime  |
realtime-dev.supabase-realtime  | 17:36:28.027 [error] Postgrex.Protocol (#PID<0.136.0>) failed to connect: ** (DBConnection.ConnectionError) tcp connect (db:5432): timeout
realtime-dev.supabase-realtime  | 17:36:28.027 [error] Postgrex.Protocol (#PID<0.135.0>) failed to connect: ** (DBConnection.ConnectionError) tcp connect (db:5432): timeout
realtime-dev.supabase-realtime  | ** (DBConnection.ConnectionError) connection not available and request was dropped from queue after 14870ms. This means requests are coming in and your connection pool cannot serve them fast enough. You can address this by:
realtime-dev.supabase-realtime  |
realtime-dev.supabase-realtime  |   1. Ensuring your database is available and that you can connect to it

realtime-dev.supabase-realtime  |   2. Tracking down slow queries and making sure they are running fast enough
realtime-dev.supabase-realtime  |   3. Increasing the pool_size (although this increases resource consumption)
realtime-dev.supabase-realtime  |   4. Allowing requests to wait longer by increasing :queue_target and :queue_interval
realtime-dev.supabase-realtime  |
realtime-dev.supabase-realtime  | See DBConnection.start_link/2 for more information
realtime-dev.supabase-realtime  |
realtime-dev.supabase-realtime  |     (ecto_sql 3.8.3) lib/ecto/adapters/sql.ex:932: Ecto.Adapters.SQL.raise_sql_call_error/1
realtime-dev.supabase-realtime  |     (elixir 1.14.0) lib/enum.ex:1658: Enum."-map/2-lists^map/1-0-"/2
realtime-dev.supabase-realtime  |     (ecto_sql 3.8.3) lib/ecto/adapters/sql.ex:1024: Ecto.Adapters.SQL.execute_ddl/4
realtime-dev.supabase-realtime  |     (ecto_sql 3.8.3) lib/ecto/migrator.ex:696: Ecto.Migrator.verbose_schema_migration/3
realtime-dev.supabase-realtime  |     (ecto_sql 3.8.3) lib/ecto/migrator.ex:510: Ecto.Migrator.lock_for_migrations/4
realtime-dev.supabase-realtime  |     (ecto_sql 3.8.3) lib/ecto/migrator.ex:422: Ecto.Migrator.run/4
realtime-dev.supabase-realtime  |     (ecto_sql 3.8.3) lib/ecto/migrator.ex:146: Ecto.Migrator.with_repo/3
realtime-dev.supabase-realtime  |     nofile:1: (file)
realtime-dev.supabase-realtime exited with code 0
supabase-storage                | Error: timeout expired
supabase-storage                |     at Timeout._onTimeout (/app/node_modules/pg/lib/client.js:106:28)
supabase-storage                |     at listOnTimeout (node:internal/timers:564:17)
supabase-storage                |     at processTimers (node:internal/timers:507:7)
```
